PerfTools/AllocMonitor/plugins/ModuleEventAllocMonitor.cc

```diff
@@ -0,0 +1,445 @@
+      static ThreadAllocInfo s_info[allocMon::ThreadTracker::kTotalEntries];
+      return s_info[allocMon::ThreadTracker::instance().thread_index()];
+      ;
+    }
+    void allocCalled(size_t iRequested, size_t iActual, void const* iAddress) final {
+      auto& allocInfo = threadAllocInfo();
+      if (not allocInfo.active_) {
+        return;
+      }
+      allocInfo.alloc(iAddress, iActual);
+    }
+    void deallocCalled(size_t iActual, void const* iAddress) final {
+      auto& allocInfo = threadAllocInfo();
+      if (not allocInfo.active_) {
+        return;
+      }
+
+      allocInfo.dealloc(iAddress, iActual);
+    }
+  };
+
+  class Filter {
+  public:
+    //a negative module id corresponds to an ES module
+    Filter(std::vector<int> const* moduleIDs);
+    //returns true if should keep this
+    bool startOnThread(int moduleID) const;
+    bool startOnThread(int moduleID, const AllocMap&) const;
+    const ThreadAllocInfo* stopOnThread(int moduleID) const;
+
+    bool startOnThread() const;
+    const ThreadAllocInfo* stopOnThread() const;
+
+    void setGlobalKeep(bool iShouldKeep);
+    bool globalKeep() const { return globalKeep_.load(); }
+
+    bool keepModuleInfo(int moduleID) const;
+
+  private:
+    std::atomic<bool> globalKeep_ = true;
+    std::vector<int> const* moduleIDs_ = nullptr;
+  };
+
+  Filter::Filter(std::vector<int> const* moduleIDs) : moduleIDs_{moduleIDs} {}
+
+  bool Filter::startOnThread(int moduleID) const {
+    if (not globalKeep_.load()) {
+      return false;
+    }
+    if (keepModuleInfo(moduleID)) {
+      MonitorAdaptor::startOnThread();
+      return true;
+    }
+    return false;
+  }
+
+  bool Filter::startOnThread(int moduleID, AllocMap const& iInfo) const {
+    if (not globalKeep_.load()) {
+      return false;
+    }
+    if (keepModuleInfo(moduleID)) {
+      MonitorAdaptor::startOnThread(iInfo);
+      return true;
+    }
+    return false;
+  }
+
+  const ThreadAllocInfo* Filter::stopOnThread(int moduleID) const {
+    if (not globalKeep_.load()) {
+      return nullptr;
+    }
+
+    if (keepModuleInfo(moduleID)) {
+      return &MonitorAdaptor::stopOnThread();
+    }
+    return nullptr;
+  }
+
+  bool Filter::startOnThread() const {
+    if (not globalKeep_.load()) {
+      return false;
+    }
+    MonitorAdaptor::startOnThread();
+    return true;
+  }
+
+  const ThreadAllocInfo* Filter::stopOnThread() const {
+    if (not globalKeep_.load()) {
+      return nullptr;
+    }
+    return &MonitorAdaptor::stopOnThread();
+  }
+
+  void Filter::setGlobalKeep(bool iShouldKeep) { globalKeep_.store(iShouldKeep); }
+
+  bool Filter::keepModuleInfo(int moduleID) const {
+    if ((nullptr == moduleIDs_) or (moduleIDs_->empty()) or
+        (std::binary_search(moduleIDs_->begin(), moduleIDs_->end(), moduleID))) {
+      return true;
+    }
+    return false;
+  }
+}  // namespace
+
+class ModuleEventAllocMonitor {
+  unsigned int moduleIndex(unsigned int mod_id) const {
```

Comment:
Would it make sense to move this function to the `private` section at the end of the class definition?

Reply:
done